aten/src/ATen/native/mkldnn/RNN.cpp

```diff
@@ -0,0 +1,583 @@
+      "get_mkldnn_tensor expects CPU tensor input");
+  TORCH_CHECK(
+      tensor.layout() == at::Layout::Strided,
+      "get_mkldnn_tensor expects dense tensor input");
+  TORCH_CHECK(
+      tensor.scalar_type() == at::ScalarType::Float ||
+          tensor.scalar_type() == at::ScalarType::BFloat16 ||
+          tensor.scalar_type() == at::ScalarType::Half,
+      "get_mkldnn_tensor expects float or bfloat16 tensor input");
+  return {desc, tensor.data_ptr()};
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor> mkldnn_rnn_layer(const Tensor& input,
+    const Tensor& w0,
+    const Tensor& w1,
+    const Tensor& w2,
+    const Tensor& w3,
+    const Tensor& hx_,
+    const Tensor& cx_,
+    bool reverse,
+    IntArrayRef batch_sizes,
+    int64_t mode,
+    int64_t hidden_size,
+    int64_t num_layers,
+    bool has_biases,
+    bool bidirectional,
+    bool batch_first,
+    bool train) {
+  RNNParams rnn(
+      input,
+      batch_sizes,
+      mode,
+      hidden_size,
+      num_layers,
+      bidirectional,
+      batch_first,
+      train);
+
+  auto output_size = _output_size</*is_single_direction*/ true>(rnn);
+  auto output = at::empty(output_size, input.options());
+
+  auto hy_ = at::empty(hx_.sizes(), hx_.options());
+  auto cy_ = at::empty(cx_.sizes(), cx_.options());
+
+  auto weight_ih = _shuffle_weight(w0, rnn.mode);
+  auto weight_hh = _shuffle_weight(w1, rnn.mode);
+
+  auto bias = has_biases
+      ? _shuffle_bias(w2, w3, rnn.mode)
+      : at::zeros({rnn.num_bias_gates * rnn.hidden_size}, weight_ih.options());
+
+  // per layer input size
+  int64_t input_size = input.size(2);
+  auto x = get_mkldnn_tensor(
+      input,
+      rnn.src_layer_desc(input_size, get_mkldnn_dtype(input.scalar_type())));
+  auto hx = get_mkldnn_tensor(
+      hx_, rnn.src_iter_desc(get_mkldnn_dtype(hx_.scalar_type())));
+  auto cx = get_mkldnn_tensor(
+      cx_, rnn.src_iter_c_desc(get_mkldnn_dtype(cx_.scalar_type())));
+  auto b = get_mkldnn_tensor(
+      bias, rnn.bias_desc(get_mkldnn_dtype(bias.scalar_type())));
+  auto y = get_mkldnn_tensor(
+      output, rnn.dst_layer_desc(get_mkldnn_dtype(output.scalar_type())));
+  auto hy = get_mkldnn_tensor(
+      hy_, rnn.dst_iter_desc(get_mkldnn_dtype(hy_.scalar_type())));
+  auto cy = get_mkldnn_tensor(
+      cy_, rnn.dst_iter_c_desc(get_mkldnn_dtype(cy_.scalar_type())));
+  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih.scalar_type())));
+  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh.scalar_type())));
+
+  Tensor workspace = Tensor();
+  auto pd = ideep::lstm_forward_training::prepare(
+      x, hx, cx, w1_, w2_, b, y, hy, cy, reverse);
+  workspace = at::empty(pd.workspace_desc().get_size() / sizeof(uint8_t), input.options().dtype(at::kByte));
+  ideep::tensor mkldnn_workspace;
+  mkldnn_workspace.init(
+      pd.workspace_desc(), workspace.template data_ptr<uint8_t>());
```

Comment:
Why `template` keyword is needed here?
```suggestion
      pd.workspace_desc(), workspace.data_ptr<uint8_t>());
```